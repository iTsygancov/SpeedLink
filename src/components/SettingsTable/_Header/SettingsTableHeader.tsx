import { headerColumns } from "./SettingsTableHeader.data";
import { TableHead, TableHeader, TableRow } from "@/components/ui/table";
import { SortBy, SortByColumn } from "@/types";
import { ArrowDownNarrowWide, ArrowUpWideNarrow } from "lucide-react";

export type SettingsTableHeaderProps = {
  handleSort: (column: SortByColumn) => void;
  sortBy: SortBy;
};

const SettingsTableHeader = ({
  handleSort,
  sortBy
}: SettingsTableHeaderProps) => {
  const renderSortIndicator = (column: SortByColumn) => {
    const isAscending = sortBy.column === column && sortBy.direction === "asc";
    const ArrowIcon = isAscending ? ArrowDownNarrowWide : ArrowUpWideNarrow;

    return <ArrowIcon size={16} />;
  };

  const renderSortableHeader = (
    column: SortByColumn,
    label: string,
    cellWidth: string
  ) => (
    <TableHead className={cellWidth} key={label}>
      <div className='flex items-center gap-2'>
        <span
          className='cursor-pointer select-none hover:underline'
          onClick={() => handleSort(column)}
        >
          {label}
        </span>
        {sortBy.column === column && renderSortIndicator(column)}
      </div>
    </TableHead>
  );

  return (
    <TableHeader>
      <TableRow>
        {headerColumns.map(({ column, label, width }) =>
          renderSortableHeader(column, label, width)
        )}
        <TableHead className='w-[350px]' />
      </TableRow>
    </TableHeader>
  );
};

export default SettingsTableHeader;
