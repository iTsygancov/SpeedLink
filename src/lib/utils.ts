import { type ClassValue, clsx } from "clsx";
import { twMerge } from "tailwind-merge";

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function generateAlphanumericKeysArray() {
  const keysArray = [];

  // Generate lowercase alphabetic keys
  for (let i = 65; i <= 90; i++) {
    keysArray.push(String.fromCharCode(i));
  }

  // Generate numeric keys
  for (let i = 48; i <= 57; i++) {
    keysArray.push(String.fromCharCode(i));
  }

  return keysArray;
}
