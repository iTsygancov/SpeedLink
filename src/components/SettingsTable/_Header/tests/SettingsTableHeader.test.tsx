import SettingsTableHeader, {
  SettingsTableHeaderProps
} from "../SettingsTableHeader";
import { expectToMatchSnapshot, renderTree } from "@/test/utils";
import { describe, it, vi } from "vitest";

const initialProps: SettingsTableHeaderProps = {
  handleSort: vi.fn(),
  sortBy: { column: "shortcut", direction: "asc" }
};

const testComponentWithProps = (props: SettingsTableHeaderProps) => {
  expectToMatchSnapshot(renderTree(<SettingsTableHeader {...props} />));
};

describe("SettingsTableHeader", () => {
  it("should render with default props", () => {
    testComponentWithProps(initialProps);
  });

  it("should render with sort by shortcut in desc direction", () => {
    testComponentWithProps({
      ...initialProps,
      sortBy: {
        ...initialProps.sortBy,
        direction: "desc"
      }
    });
  });

  it("should render with sort by url in asc direction", () => {
    testComponentWithProps({
      ...initialProps,
      sortBy: {
        column: "url",
        direction: "asc"
      }
    });
  });

  it("should render with sort by url in desc direction", () => {
    testComponentWithProps({
      ...initialProps,
      sortBy: {
        column: "url",
        direction: "desc"
      }
    });
  });

  it("should render with sort by title in asc direction", () => {
    testComponentWithProps({
      ...initialProps,
      sortBy: {
        column: "title",
        direction: "asc"
      }
    });
  });

  it("should render with sort by title in desc direction", () => {
    testComponentWithProps({
      ...initialProps,
      sortBy: {
        column: "title",
        direction: "desc"
      }
    });
  });
});
