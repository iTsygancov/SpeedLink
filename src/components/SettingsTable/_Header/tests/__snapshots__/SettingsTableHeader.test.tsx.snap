// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SettingsTableHeader > should render with default props 1`] = `
<thead
  className="[&_tr]:border-b"
>
  <tr
    className="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
  >
    <th
      className="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 min-w-[200px]"
    >
      <div
        className="flex items-center gap-2"
      >
        <span
          className="cursor-pointer select-none hover:underline"
          onClick={[Function]}
        >
          Shortcut
        </span>
        <svg
          className="lucide lucide-arrow-down-narrow-wide"
          fill="none"
          height={16}
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth={2}
          viewBox="0 0 24 24"
          width={16}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m3 16 4 4 4-4"
          />
          <path
            d="M7 20V4"
          />
          <path
            d="M11 4h4"
          />
          <path
            d="M11 8h7"
          />
          <path
            d="M11 12h10"
          />
        </svg>
      </div>
    </th>
    <th
      className="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 min-w-[500px]"
    >
      <div
        className="flex items-center gap-2"
      >
        <span
          className="cursor-pointer select-none hover:underline"
          onClick={[Function]}
        >
          URL
        </span>
      </div>
    </th>
    <th
      className="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 min-w-[350px]"
    >
      <div
        className="flex items-center gap-2"
      >
        <span
          className="cursor-pointer select-none hover:underline"
          onClick={[Function]}
        >
          Title
        </span>
      </div>
    </th>
    <th
      className="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 w-[350px]"
    />
  </tr>
</thead>
`;

exports[`SettingsTableHeader > should render with sort by shortcut in desc direction 1`] = `
<thead
  className="[&_tr]:border-b"
>
  <tr
    className="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
  >
    <th
      className="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 min-w-[200px]"
    >
      <div
        className="flex items-center gap-2"
      >
        <span
          className="cursor-pointer select-none hover:underline"
          onClick={[Function]}
        >
          Shortcut
        </span>
        <svg
          className="lucide lucide-arrow-up-wide-narrow"
          fill="none"
          height={16}
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth={2}
          viewBox="0 0 24 24"
          width={16}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m3 8 4-4 4 4"
          />
          <path
            d="M7 4v16"
          />
          <path
            d="M11 12h10"
          />
          <path
            d="M11 16h7"
          />
          <path
            d="M11 20h4"
          />
        </svg>
      </div>
    </th>
    <th
      className="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 min-w-[500px]"
    >
      <div
        className="flex items-center gap-2"
      >
        <span
          className="cursor-pointer select-none hover:underline"
          onClick={[Function]}
        >
          URL
        </span>
      </div>
    </th>
    <th
      className="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 min-w-[350px]"
    >
      <div
        className="flex items-center gap-2"
      >
        <span
          className="cursor-pointer select-none hover:underline"
          onClick={[Function]}
        >
          Title
        </span>
      </div>
    </th>
    <th
      className="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 w-[350px]"
    />
  </tr>
</thead>
`;

exports[`SettingsTableHeader > should render with sort by title in asc direction 1`] = `
<thead
  className="[&_tr]:border-b"
>
  <tr
    className="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
  >
    <th
      className="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 min-w-[200px]"
    >
      <div
        className="flex items-center gap-2"
      >
        <span
          className="cursor-pointer select-none hover:underline"
          onClick={[Function]}
        >
          Shortcut
        </span>
      </div>
    </th>
    <th
      className="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 min-w-[500px]"
    >
      <div
        className="flex items-center gap-2"
      >
        <span
          className="cursor-pointer select-none hover:underline"
          onClick={[Function]}
        >
          URL
        </span>
      </div>
    </th>
    <th
      className="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 min-w-[350px]"
    >
      <div
        className="flex items-center gap-2"
      >
        <span
          className="cursor-pointer select-none hover:underline"
          onClick={[Function]}
        >
          Title
        </span>
        <svg
          className="lucide lucide-arrow-down-narrow-wide"
          fill="none"
          height={16}
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth={2}
          viewBox="0 0 24 24"
          width={16}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m3 16 4 4 4-4"
          />
          <path
            d="M7 20V4"
          />
          <path
            d="M11 4h4"
          />
          <path
            d="M11 8h7"
          />
          <path
            d="M11 12h10"
          />
        </svg>
      </div>
    </th>
    <th
      className="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 w-[350px]"
    />
  </tr>
</thead>
`;

exports[`SettingsTableHeader > should render with sort by title in desc direction 1`] = `
<thead
  className="[&_tr]:border-b"
>
  <tr
    className="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
  >
    <th
      className="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 min-w-[200px]"
    >
      <div
        className="flex items-center gap-2"
      >
        <span
          className="cursor-pointer select-none hover:underline"
          onClick={[Function]}
        >
          Shortcut
        </span>
      </div>
    </th>
    <th
      className="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 min-w-[500px]"
    >
      <div
        className="flex items-center gap-2"
      >
        <span
          className="cursor-pointer select-none hover:underline"
          onClick={[Function]}
        >
          URL
        </span>
      </div>
    </th>
    <th
      className="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 min-w-[350px]"
    >
      <div
        className="flex items-center gap-2"
      >
        <span
          className="cursor-pointer select-none hover:underline"
          onClick={[Function]}
        >
          Title
        </span>
        <svg
          className="lucide lucide-arrow-up-wide-narrow"
          fill="none"
          height={16}
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth={2}
          viewBox="0 0 24 24"
          width={16}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m3 8 4-4 4 4"
          />
          <path
            d="M7 4v16"
          />
          <path
            d="M11 12h10"
          />
          <path
            d="M11 16h7"
          />
          <path
            d="M11 20h4"
          />
        </svg>
      </div>
    </th>
    <th
      className="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 w-[350px]"
    />
  </tr>
</thead>
`;

exports[`SettingsTableHeader > should render with sort by url in asc direction 1`] = `
<thead
  className="[&_tr]:border-b"
>
  <tr
    className="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
  >
    <th
      className="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 min-w-[200px]"
    >
      <div
        className="flex items-center gap-2"
      >
        <span
          className="cursor-pointer select-none hover:underline"
          onClick={[Function]}
        >
          Shortcut
        </span>
      </div>
    </th>
    <th
      className="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 min-w-[500px]"
    >
      <div
        className="flex items-center gap-2"
      >
        <span
          className="cursor-pointer select-none hover:underline"
          onClick={[Function]}
        >
          URL
        </span>
        <svg
          className="lucide lucide-arrow-down-narrow-wide"
          fill="none"
          height={16}
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth={2}
          viewBox="0 0 24 24"
          width={16}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m3 16 4 4 4-4"
          />
          <path
            d="M7 20V4"
          />
          <path
            d="M11 4h4"
          />
          <path
            d="M11 8h7"
          />
          <path
            d="M11 12h10"
          />
        </svg>
      </div>
    </th>
    <th
      className="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 min-w-[350px]"
    >
      <div
        className="flex items-center gap-2"
      >
        <span
          className="cursor-pointer select-none hover:underline"
          onClick={[Function]}
        >
          Title
        </span>
      </div>
    </th>
    <th
      className="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 w-[350px]"
    />
  </tr>
</thead>
`;

exports[`SettingsTableHeader > should render with sort by url in desc direction 1`] = `
<thead
  className="[&_tr]:border-b"
>
  <tr
    className="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
  >
    <th
      className="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 min-w-[200px]"
    >
      <div
        className="flex items-center gap-2"
      >
        <span
          className="cursor-pointer select-none hover:underline"
          onClick={[Function]}
        >
          Shortcut
        </span>
      </div>
    </th>
    <th
      className="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 min-w-[500px]"
    >
      <div
        className="flex items-center gap-2"
      >
        <span
          className="cursor-pointer select-none hover:underline"
          onClick={[Function]}
        >
          URL
        </span>
        <svg
          className="lucide lucide-arrow-up-wide-narrow"
          fill="none"
          height={16}
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth={2}
          viewBox="0 0 24 24"
          width={16}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m3 8 4-4 4 4"
          />
          <path
            d="M7 4v16"
          />
          <path
            d="M11 12h10"
          />
          <path
            d="M11 16h7"
          />
          <path
            d="M11 20h4"
          />
        </svg>
      </div>
    </th>
    <th
      className="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 min-w-[350px]"
    >
      <div
        className="flex items-center gap-2"
      >
        <span
          className="cursor-pointer select-none hover:underline"
          onClick={[Function]}
        >
          Title
        </span>
      </div>
    </th>
    <th
      className="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 w-[350px]"
    />
  </tr>
</thead>
`;
